// sanity-plugin-order-documents/build/components/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

var _immutabilityHelper = _interopRequireDefault(require("immutability-helper"));

var _reactSelect = _interopRequireDefault(require("react-select"));

var _reactDnd = require("react-dnd");

var _reactDndHtml5Backend = require("react-dnd-html5-backend");

var _spinner = _interopRequireDefault(require("part:@sanity/components/loading/spinner"));

var _preview = _interopRequireDefault(require("part:@sanity/base/preview"));

var _client = _interopRequireDefault(require("part:@sanity/base/client"));

var _schema = _interopRequireDefault(require("part:@sanity/base/schema"));

var _router = require("part:@sanity/base/router");

var _index = _interopRequireDefault(require("../index.css"));

var _functions = require("../functions");

var _Card = require("./Card");

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2["default"])(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2["default"])(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2["default"])(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

var OrderDocuments = /*#__PURE__*/function (_React$Component) {
  (0, _inherits2["default"])(OrderDocuments, _React$Component);

  var _super = _createSuper(OrderDocuments);

  function OrderDocuments() {
    var _this;

    (0, _classCallCheck2["default"])(this, OrderDocuments);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "state", {
      documents: [],
      type: ""
    });
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "observables", {});
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "handleReceiveList", /*#__PURE__*/function () {
      var _ref = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee(documents) {
        return _regenerator["default"].wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                _this.setState({
                  documents: documents
                });

                _context.next = 3;
                return (0, _functions.setListOrder)(documents);

              case 3:
              case "end":
                return _context.stop();
            }
          }
        }, _callee);
      }));

      return function (_x) {
        return _ref.apply(this, arguments);
      };
    }());
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "handleChange", function (_ref2) {
      var value = _ref2.value;

      _this.setState({
        type: value
      });

      _this.observables = _client["default"].observable.fetch('*[!(_id in path("drafts.**")) && _type == $types][0...100] | order (order asc, _updatedAt desc)', {
        types: value
      }).subscribe(_this.handleReceiveList);
    });
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "moveCard", /*#__PURE__*/function () {
      var _ref3 = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee2(beforeIndex, afterIndex) {
        var card1, card2;
        return _regenerator["default"].wrap(function _callee2$(_context2) {
          while (1) {
            switch (_context2.prev = _context2.next) {
              case 0:
                card1 = _this.state.documents[beforeIndex];
                card2 = _this.state.documents[afterIndex];

                _this.setState({
                  documents: (0, _immutabilityHelper["default"])(_this.state.documents, {
                    $splice: [[beforeIndex, 1], [afterIndex, 0, card1]]
                  })
                });

                _context2.next = 5;
                return Promise.all([(0, _functions.setOrder)(card1._id, afterIndex), (0, _functions.setOrder)(card2._id, beforeIndex)]);

              case 5:
              case "end":
                return _context2.stop();
            }
          }
        }, _callee2);
      }));

      return function (_x2, _x3) {
        return _ref3.apply(this, arguments);
      };
    }());
    return _this;
  }

  (0, _createClass2["default"])(OrderDocuments, [{
    key: "renderDraggableSection",
    value: function renderDraggableSection() {
      var _this2 = this;

      var _this$state = this.state,
          documents = _this$state.documents,
          type = _this$state.type;

      if (!type && !documents.length) {
        return null;
      }

      if (type && !documents.length) {
        return /*#__PURE__*/_react["default"].createElement("div", {
          className: _index["default"].marginTop
        }, /*#__PURE__*/_react["default"].createElement(_spinner["default"], {
          message: "Loading...",
          center: true
        }));
      }

      return /*#__PURE__*/_react["default"].createElement(_react["default"].Fragment, null, /*#__PURE__*/_react["default"].createElement("hr", {
        className: _index["default"].rule
      }), /*#__PURE__*/_react["default"].createElement("p", null, /*#__PURE__*/_react["default"].createElement("strong", null, "Step 2: Drag and Drop to Re-order")), /*#__PURE__*/_react["default"].createElement("ul", {
        className: _index["default"].list
      }, documents.map(function (document, index) {
        return /*#__PURE__*/_react["default"].createElement("li", {
          key: document._id,
          className: _index["default"].listItem
        }, /*#__PURE__*/_react["default"].createElement(_Card.Card, {
          key: document._id,
          index: index,
          id: document._id,
          text: document.title,
          moveCard: _this2.moveCard,
          jsx: /*#__PURE__*/_react["default"].createElement(_preview["default"], {
            value: document,
            type: _schema["default"].get(document._type)
          })
        }));
      })));
    }
  }, {
    key: "renderDocumentsList",
    value: function renderDocumentsList() {
      var documents = this.state.documents;

      if (!documents) {
        return /*#__PURE__*/_react["default"].createElement("div", {
          className: _index["default"].list
        }, /*#__PURE__*/_react["default"].createElement(_spinner["default"], {
          message: "Loading...",
          center: true
        }));
      }

      var uniqueTypes = ((0, _functions.getDocumentTypeNames)() || []).map(function (_ref4) {
        var name = _ref4.name,
            title = _ref4.title;
        return {
          value: name,
          label: title
        };
      });
      return /*#__PURE__*/_react["default"].createElement(_react["default"].Fragment, null, /*#__PURE__*/_react["default"].createElement("h2", {
        className: _index["default"].noTopMargin
      }, "Order Documents"), /*#__PURE__*/_react["default"].createElement("p", null, "Order your documents via drag-and-drop."), /*#__PURE__*/_react["default"].createElement("hr", null), /*#__PURE__*/_react["default"].createElement("p", null, /*#__PURE__*/_react["default"].createElement("strong", null, "Step 1: Choose a Type")), /*#__PURE__*/_react["default"].createElement(_reactSelect["default"], {
        options: uniqueTypes,
        isSearchable: true,
        onChange: this.handleChange
      }));
    }
  }, {
    key: "render",
    value: function render() {
      var _this$state2 = this.state,
          documents = _this$state2.documents,
          type = _this$state2.type;
      return /*#__PURE__*/_react["default"].createElement(_reactDnd.DndProvider, {
        backend: _reactDndHtml5Backend.HTML5Backend
      }, /*#__PURE__*/_react["default"].createElement("div", {
        className: _index["default"].container
      }, /*#__PURE__*/_react["default"].createElement("div", {
        className: _index["default"].outerWrapper
      }, /*#__PURE__*/_react["default"].createElement("div", {
        className: _index["default"].innerWrapper
      }, this.renderDocumentsList(), this.renderDraggableSection()))));
    }
  }]);
  return OrderDocuments;
}(_react["default"].Component);

var _default = (0, _router.withRouterHOC)(MyTool);

exports["default"] = _default;
