// sanity-plugin-order-documents/build/components/TypeSection.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _react = _interopRequireDefault(require("react"));

var _reactSelect = _interopRequireDefault(require("react-select"));

var _data = require("../data");

var _functions = require("../functions");

var _index = _interopRequireDefault(require("../index.css"));

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2["default"])(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2["default"])(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2["default"])(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

var TypeSection = /*#__PURE__*/function (_React$Component) {
  (0, _inherits2["default"])(TypeSection, _React$Component);

  var _super = _createSuper(TypeSection);

  function TypeSection() {
    (0, _classCallCheck2["default"])(this, TypeSection);
    return _super.apply(this, arguments);
  }

  (0, _createClass2["default"])(TypeSection, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
          documents = _this$props.documents,
          type = _this$props.type,
          field = _this$props.field,
          fields = _this$props.fields,
          showFields = _this$props.showFields,
          handleChange = _this$props.handleChange,
          handleFieldChange = _this$props.handleFieldChange;

      if (!documents) {
        return /*#__PURE__*/_react["default"].createElement("div", {
          className: _index["default"].list
        }, /*#__PURE__*/_react["default"].createElement(Spinner, {
          message: "Loading...",
          center: true
        }));
      }

      var uniqueTypes = ((0, _functions.getDocumentTypeNames)(field) || []).map(function (_ref) {
        var name = _ref.name,
            title = _ref.title;
        return {
          value: name,
          label: title
        };
      });
      var uniqueFields = fields.map(function (_ref2) {
        var name = _ref2.name,
            title = _ref2.title;
        return {
          value: name,
          label: name === _data.DEFAULT_FIELD_VALUE ? _data.DEFAULT_FIELD_LABEL : title
        };
      });
      return /*#__PURE__*/_react["default"].createElement(_react["default"].Fragment, null, /*#__PURE__*/_react["default"].createElement("div", {
        className: _index["default"].flexSpaceBetween
      }, /*#__PURE__*/_react["default"].createElement("div", null, /*#__PURE__*/_react["default"].createElement("h2", {
        className: _index["default"].noTopMargin
      }, "Order Documents"), /*#__PURE__*/_react["default"].createElement("p", null, "Order your documents via drag-and-drop.")), /*#__PURE__*/_react["default"].createElement("div", null, showFields ? /*#__PURE__*/_react["default"].createElement("div", {
        className: _index["default"].fieldButton
      }, /*#__PURE__*/_react["default"].createElement(_reactSelect["default"], {
        options: uniqueFields,
        isSearchable: true,
        onChange: this.handleFieldChange,
        defaultValue: {
          value: _data.DEFAULT_FIELD_VALUE,
          label: _data.DEFAULT_FIELD_LABEL
        }
      })) : null)), /*#__PURE__*/_react["default"].createElement("hr", null), /*#__PURE__*/_react["default"].createElement("p", null, /*#__PURE__*/_react["default"].createElement("strong", null, "Step 1: Choose a Type")), /*#__PURE__*/_react["default"].createElement(_reactSelect["default"], {
        options: uniqueTypes,
        isSearchable: true,
        onChange: handleChange,
        value: type
      }));
    }
  }]);
  return TypeSection;
}(_react["default"].Component);

var _default = TypeSection;
exports["default"] = _default;
